### PR TITLE
Implement HMAC validator

### PR DESCRIPTION
**Pull request by sndrgrdn**
This will add functionality to verify HMAC signatures from the webhooks that are received.

There are specs available for this new behaviour.

### CODE REVIEW

**Comment by crrood** [comment]
Hi Sander,

Thanks for adding this functionality, it looks good to me.

Have you tested it in the sandbox environment, using a real-time HMAC signature?

Best,
Colin

**Comment by sndrgrdn** [comment]
> Have you tested it in the sandbox environment, using a real-time HMAC signature?

I'm currently implementing Adyen into our webshop and needed it to verify webhooks. so yes this is being testing in the sandbox environment with real-time HMAC signatures :)

To give an example how it is being used in a Rails project, I included a snippet below:

```ruby
hmac_key = 'YOUR_HMAC_KEY'
validator = Adyen::Utils::HmacValidator.new

params['notificationItems'].each do |notification_item|
  request_item = notification_item['NotificationRequestItem']
  next unless validator.validate(request_item, hmac_key)

  # Call some worker or other business logic here
end

render :text => '[accepted]'

```

**Comment by crrood** [comment]
Hi Sander,

Okay, sounds good - I've approved the PR, just need one of my colleagues to approve as well and we'll merge it.

Best,
Colin

File: lib/adyen/utils/hmac_validator.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  module Utils
+    class HmacValidator
+      HMAC_ALGORITHM = 'sha256'.freeze
+      DATA_SEPARATOR = ':'.freeze
+      VALIDATION_KEYS = %w[
+        pspReference originalReference merchantAccountCode merchantReference
+        amount.value amount.currency eventCode success
+      ].freeze
+
+      def validate(notification_request_item, hmac_key)
```
Comment (rikterbeek) [inline lib/adyen/utils/hmac_validator.rb]: Could you rename this to is_valid_notification_hmac ?

File: lib/adyen/utils/hmac_validator.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        expected_sign = calculate_hmac(notification_request_item, hmac_key)
+        merchant_sign = fetch(notification_request_item, 'additionalData.hmacSignature')
+
+        expected_sign == merchant_sign
+      end
+
+      def calculate_hmac(notification_request_item, hmac_key)
```
Comment (rikterbeek) [inline lib/adyen/utils/hmac_validator.rb]: Could you rename this to  calculate_notification_hmac ?

File: lib/adyen/utils/hmac_validator.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
+module Adyen
+  module Utils
+    class HmacValidator
+      HMAC_ALGORITHM = 'sha256'.freeze
+      DATA_SEPARATOR = ':'.freeze
+      VALIDATION_KEYS = %w[
```
Comment (rikterbeek) [inline lib/adyen/utils/hmac_validator.rb]: Could you rename this to NOTIFICATION_VALIDATION_KEYS ?

**Comment by rikterbeek** [comment]
Hi @sndrgrdn,

Many thanks for this PR!. Definitely we wanted to add so great that you have done a PR for this.
As we are wanting to make this feature similar across all libraries could you adjust the naming of the methods ? For the rest it looks great :)

Regards,
Rik
Adyen

File: lib/adyen/utils/hmac_validator.rb (hunk 1 above)
Comment (sndrgrdn) [reply lib/adyen/utils/hmac_validator.rb]: If you don't mind, I would like to suggest naming it to `valid_notification_hmac?` instead, this is the preferred way to define methods that return booleans.

ref: https://rubystyle.guide/#bool-methods-qmark

**Pull request merged**